### DISCUSSION TITLE
Vendor-neutral open-source Chrome Webstore alternative?

### DISCUSSION THREAD

**Discussion by vairag22** [Eloston/ungoogled-chromium#1559] [Q&A]
Are there any vendor-neutral open-source alternative to the Chrome Webstore? You know like [open-vsx.org](https://open-vsx.org), an alternative to Visual Studio Marketplace. Did anyone ever try to do it?

**Comment by megastallman** [comment]
I was also thinking about that kind of option, but it probably needs much bigger effort from UC team, and it would be too difficult to catch up with google market. Thinking even, not about open-vsx.org, but more about F-Droid store.
But I basically think that google market is not that dangerous. It is more about extension developrs - to avoid adding telemetry and other trojan functional.

One more idea - why not adding https://github.com/NeverDecaf/chromium-web-store during UC build process? But this Chromium Web Store plugin periodically connects to 2 google domains to check/download plugins. So, I keep it disabled all the time, because I prefer a silent browser, that doesn't knock everywhere when not asked to.

So, maybe the guys are right about providing us the best of that browser, and then we can compromise it as much as we want.

**Reply by wchen342** [reply]
I think this is actually a good point. I actually wonder why it is not included. @PF4Public @Eloston @Ahrotahn is there a specific reason `chromium-web-store` is not included by default?

**Reply by PF4Public** [reply]
I may name some of the reasons off the top of my head:
* `chromium-web-store` is out of control of ungoogled-chromium, which makes its inclusion a dubious decision
* Given the point above its installation should be a well-weighed decision of a end-user oneself
* I bet there are users, who see "ungoogled" and don't expect to see some interaction (albeit minor) with google servers. Lastly, should they remove the extension upon every update?
* It is relatively easy to install it. And it needs to be done only once for a profile. Later it will update itself.

In other words. Imagine that I want to compile ungoogled-chromium. W.r.t. `chromium-web-store` I have two options: embed it at the build-time and install later as an extension. I think that embedding is not a trivial task. Considering all the points above I could conclude that installing `chromium-web-store` into profile as a normal extension instead of embedding it outweights.

**Reply by wchen342** [reply]
I have discussed this with @networkException in another thread but basically,

> chromium-web-store is out of control of ungoogled-chromium, which makes its inclusion a dubious decision

We can easily fork it. The extension is not complicated and doesn't update often, so it is easy to maintain.

> Given the point above its installation should be a well-weighed decision of a end-user oneself

This is true. However there are also users expecting being able to install extensions as part of the functionality of chromium, and the instruction that directs the user to wiki is not explicit if the user doesn't know where to look at. Maybe we shall at least add something to the extension page pointing to the wiki link.

> I bet there are users, who see "ungoogled" and don't expect to see some interaction (albeit minor) with google servers. Lastly, should they remove the extension upon every update?

The update checking that pings google server can be disabled in the extension, otherwise it won't activate unless the user manually visits the web store page. I am not sure I understood the "remove on every update" part. Is it because they want to "ungoogled" everything?

> I think that embedding is not a trivial task

It can be included easily. I did it for every version actually when I was still maintaining `ungoogled-chromium-android-extension` before, and I made the patch once and never needed to touch it again. My proposal in the other thread is include the extension as disabled by default, and the user can enable it if they want to use web store, otherwise left it disabled.

**Reply by networkException** [reply]
I think we need to be strict about what goes into the main project but installing extensions is clearly something in the interest of users. If we include it I would propose the following criteria:

- Make it completely optional. The extension menu is a good place but a flag would work as well
- Even when active it should not ping google for updates by default
- The code is fully present in this repository and under our control
- The addition of this is disclosed in release changelogs with reasoning and implementation details

**Reply by PF4Public** [reply]
> I am not sure I understood the "remove on every update" part. 

Is it possible to remove built-in extensions? I think I confused you (and myself) with this. I thought it would be possible to remove the extension, but it will return with every browser update. 

> Is it because they want to "ungoogled" everything?

Yep, I imagine following bug-reports: "Why are you secretly sneaking this extension into my browser? I didn't ask!" and so on. I may be just overreacting :)

> It can be included easily.

Well, you know better!

In essence my argument is: is it worth the trouble? Lets see, what others say.

**Reply by wchen342** [reply]
> "Why are you secretly sneaking this extension into my browser? I didn't ask!"

Yeah, that is actually my biggest worry about including it, thus I ever only tried it in `-android-extension` which is basically my sole responsibility as well as being marked explicitly as highly experimental. I obviously do not want to damage the name of `ungoogled-chromium` by mistake like this.

> I thought it would be possible to remove the extension, but it will return with every browser update

If it is bundled..then yes it will return. That is also a problem right.

So from the discussion I think a better solution is a better guide for users who want to install it. Someone mentioned we can add a landing page for users first using the browser. We can use a landing page to do a lot of things, including educating users on choosing functionalities they want and tell them where to find other necessary parts.

**Reply by PF4Public** [reply]
> Someone mentioned we can add a landing page for users first using the browser.

Wonderful idea!

> better solution is a better guide for users who want to install it

Could it be that main README is terrifyingly huge and could be somehow better structured for users to find needed information easier?

**Reply by wchen342** [reply]
> Could it be that main README is terrifyingly huge and could be somehow better structured for users to find needed information easier?

Yes and No. The README does need a reform since it is way too long IMO and a lot of information are buried inside huge chunk of texts. However, no README can be a substitution of hints and helps in the actual application. There are a lot of people who either get the software from somewhere else thus never reached this repo, or they simply don't have the habit of reading READMEs. Some kind of in-application help is necessary.

**Reply by Ahrotahn** [reply]
It would be pretty easy to have a page open on first launch. I think the FAQ would be a better page than the readme since the browser is already installed and end users would be more likely to want to know how to install extensions, enable spellcheck, widevine, etc.

**Reply by vairag22** [reply]
Yeah FAQ is more suitable than README. But I think a custom launch page that has both FAQ and Changelog will be better as it would cater both new and old users.

**Reply by wchen342** [reply]
Yeah I think a landing page that launches _each update_ instead of only first installation is the best option, because it counts both new and old users. We can add a switch for each page visited so old users won't keep visiting seen pages. FAQ is good but it has the same probelm with README: not all users know to visit them.

**Reply by PF4Public** [reply]
> Yeah I think a landing page that launches _each update_ instead of only first installation is the best option

Please don't do this to me! I hate it when Vivaldi wants to brag about itself upon each update! I simply do not care!

**Reply by Ahrotahn** [reply]
Concerning the original topic:

Any servers that any of us host would be just as prone to outages and compromises as Googles's servers if not more so since we're not a large corporation.  

I think if we were to go so far as to try to include the webstore extension then it would be less effort to use the existing code in Chromium by setting the URL for the webstore behind a disabled flag.  The URL's value could be contained in the flag choices header which we have full control over and isn't domain substituted since it's created at patchtime.  This would also allow end users the ability to set the URL to whichever server they are comfortable using instead of being dictated by us.  

I don't believe that undergoing such a project is something that would be all that beneficial, but only for the current state of affairs.  Once manifest v3 rolls out there may be renewed interest once devs are no longer allowed to submit v2 extensions.  I guess we'll have to see how things pan out with that.
